File: web/src/lib/shippingUtils.ts
import type { Customer, Order } from '@/types';
import { orderRevenue } from '@/types';
import { formatMoney } from './money';

export interface ShippingSnapshot {
  shipping_name: string;
  shipping_phone: string | null;
  shipping_address: string | null;
}

/** Địa chỉ / liên hệ hiển thị trên đơn — ưu tiên snapshot trên order, fallback customer (đơn cũ). */
export function resolveShippingName(order: Order, customer: Customer): string {
  return order.shipping_name?.trim() || customer.name.trim();
}

export function resolveShippingPhone(order: Order, customer: Customer): string {
  return order.shipping_phone?.trim() || customer.phone?.trim() || '';
}

export function resolveShippingAddress(order: Order, customer: Customer): string {
  return order.shipping_address?.trim() || customer.address?.trim() || '';
}

/** Snapshot lúc ghi đơn — luôn lưu giá trị hiệu lực (kể cả khi trùng khách). */
export function snapshotShipping(
  customer: Customer,
  input: {
    shipping_name?: string;
    shipping_phone?: string;
    shipping_address?: string;
  },
): ShippingSnapshot {
  const name = input.shipping_name?.trim() || customer.name.trim();
  const phone = input.shipping_phone?.trim() || customer.phone?.trim() || null;
  const address = input.shipping_address?.trim() || customer.address?.trim() || null;
  return { shipping_name: name, shipping_phone: phone, shipping_address: address };
}

export function shippingDiffersFromCustomer(order: Order, customer: Customer): boolean {
  const name = resolveShippingName(order, customer);
  const phone = resolveShippingPhone(order, customer);
  const address = resolveShippingAddress(order, customer);
  return (
    name !== customer.name.trim() ||
    phone !== (customer.phone?.trim() || '') ||
    address !== (customer.address?.trim() || '')
  );
}

export function formatShippingInfo(customer: Customer, order: Order): string {
  const name = resolveShippingName(order, customer);
  const phone = resolveShippingPhone(order, customer);
  const address = resolveShippingAddress(order, customer);
  const lines = [
    `Người nhận: ${name}`,
    phone ? `SĐT: ${phone}` : null,
    address ? `Địa chỉ: ${address}` : null,
    `Sản phẩm: ${order.product_name} × ${order.quantity}`,
    `Thành tiền: ${formatMoney(orderRevenue(order))}`,
    order.note ? `Ghi chú đơn: ${order.note}` : null,
  ].filter(Boolean);
  return lines.join('\n');
}

export function formatAddressOnly(customer: Customer, order?: Order): string {
  if (order) return resolveShippingAddress(order, customer);
  return customer.address?.trim() || '';
}

File: app/lib/utils/shipping_utils.dart
import '../models/customer.dart';
import '../models/order.dart';
import 'money.dart';

class ShippingSnapshot {
  final String shippingName;
  final String? shippingPhone;
  final String? shippingAddress;

  const ShippingSnapshot({
    required this.shippingName,
    this.shippingPhone,
    this.shippingAddress,
  });
}

String resolveShippingName(Order order, Customer customer) {
  final v = order.shippingName?.trim();
  if (v != null && v.isNotEmpty) return v;
  return customer.name.trim();
}

String resolveShippingPhone(Order order, Customer customer) {
  final v = order.shippingPhone?.trim();
  if (v != null && v.isNotEmpty) return v;
  return customer.phone?.trim() ?? '';
}

String resolveShippingAddress(Order order, Customer customer) {
  final v = order.shippingAddress?.trim();
  if (v != null && v.isNotEmpty) return v;
  return customer.address?.trim() ?? '';
}

ShippingSnapshot snapshotShipping(
  Customer customer, {
  String? shippingName,
  String? shippingPhone,
  String? shippingAddress,
}) {
  return ShippingSnapshot(
    shippingName: shippingName?.trim().isNotEmpty == true
        ? shippingName!.trim()
        : customer.name.trim(),
    shippingPhone: shippingPhone?.trim().isNotEmpty == true
        ? shippingPhone!.trim()
        : customer.phone?.trim(),
    shippingAddress: shippingAddress?.trim().isNotEmpty == true
        ? shippingAddress!.trim()
        : customer.address?.trim(),
  );
}

bool shippingDiffersFromCustomer(Order order, Customer customer) {
  return resolveShippingName(order, customer) != customer.name.trim() ||
      resolveShippingPhone(order, customer) != (customer.phone?.trim() ?? '') ||
      resolveShippingAddress(order, customer) != (customer.address?.trim() ?? '');
}

String formatShippingInfo(Customer customer, Order order) {
  final name = resolveShippingName(order, customer);
  final phone = resolveShippingPhone(order, customer);
  final address = resolveShippingAddress(order, customer);
  final lines = <String>[
    'Người nhận: $name',
    if (phone.isNotEmpty) 'SĐT: $phone',
    if (address.isNotEmpty) 'Địa chỉ: $address',
    'Sản phẩm: ${order.productName} × ${order.quantity}',
    'Thành tiền: ${formatMoney(order.revenue)}',
    if (order.note != null && order.note!.isNotEmpty) 'Ghi chú đơn: ${order.note}',
  ];
  return lines.join('\n');
}

String formatAddressOnly(Customer customer, [Order? order]) {
  if (order != null) return resolveShippingAddress(order, customer);
  return customer.address?.trim() ?? '';
}
